### ISSUE TITLE
Have a regular 'epoch' run of Python 3 on TaskCluster CI

### ISSUE THREAD

**Issue by stephenmcgruer** [web-platform-tests/wpt#22784]
As of 2020/04/08, we are approaching the point where we could have a regular epoch run of WPT that uses Python 3. Currently ~75% of top-level WPT directories are believed to run identically in Python 2 and Python 3, with the majority of the remaining blocked on the [wptserve RFC](https://github.com/web-platform-tests/rfcs/pull/49). Running regularly on TaskCluster would allow us to more easily compare to Python 2 (i.e. using wpt.fyi and cross-comparing two runs at the same SHA), as well as spot regressions (i.e. using wpt.fyi and cross-comparing consecutive runs of Python 3 at different SHAs).

Note that at the current time, the only tested Python 3 platform/browser combination is Linux and Chrome, hence the TaskCluster-focus here. Future work will expand Python 3 to Firefox, to Safari, and to Windows and MacOS, but that's out of scope right now :)

There are (at least) a few action items that would need to be addressed before we can get this going:

- [x] @ziransun - at the current point in time, is it possible to run *all* tests in Python 3 and just let some report ERROR results, or do they actually stop wptrunner execution? (If so, we will need to filter the run to just 'known safe' test directories).
- [x] @Hexcles - if we setup these as proper TaskCluster epoch runs that push to wpt.fyi, we'll probably need to tag them somehow to stop wpt.fyi thinking they're 'proper' runs. How would you suggest doing that?
- [ ] @jgraham - does TaskCluster have the capacity to add another epoch run? If so, on what frequency? My thought was to aim initially for Chrome Dev, running either daily or weekly.

I'm obviously also interested in general feedback in case anyone thinks this is not a step we should take now. (Seems safe to me, but maybe I've overlooked something :)).

**Comment by ziransun** [comment]
Hi @stephenmcgruer, so far what I did was running tests by directories. I haven't run ALL tests in one go. Having said that, all the directory runs gave report (e.g. number of checks in total, expected results, unexpected result and the list of failed tests.) rather than stopping wptrunner execution in the case of ERROR. I can do the "all in one" run during the coming weekend if it makes sense.

**Comment by stephenmcgruer** [comment]
No need to subject your laptop to that, I think. The fact that all the individual directory runs ran successfully gives a good probability that a full run would work too (and if not, we can figure it out and fix it!).

@Hexcles if we setup these as proper TaskCluster epoch runs that push to wpt.fyi, we'll probably need to tag them somehow to stop wpt.fyi thinking they're 'proper' runs, I believe. How would you suggest doing that?

**Comment by Hexcles** [comment]
Regarding wpt.fyi, we only have an additive filtering model -- we can only say include these labels, but not exclude some labels. I'd say the easiest way is to drop the `master` label on this branch.

For Taskcluster, we'd need to tweak this code to exclude this branch:
https://github.com/web-platform-tests/wpt.fyi/blob/65dc39a3818cd2f4e86e481b9fb8b62736a82fbf/api/taskcluster/webhook.go#L141

For Azure, it'd be:
https://github.com/web-platform-tests/wpt.fyi/blob/65dc39a3818cd2f4e86e481b9fb8b62736a82fbf/api/azure/webhook.go#L75

**Comment by stephenmcgruer** [comment]
@Hexcles - just so I'm clear, your suggestion would require spinning up a new epochs branch solely for Python 3?

**Comment by Hexcles** [comment]
Oh I misread the question.

Either way would work; a new epoch branch isn't required.

And on a second thought, there's a better way to drop the `master` label for these runs, which will work regardless of whether we have a new epoch branch. If we run this on Taskcluster and give the tasks a new name (e.g. `wpt-python3-N`), we can filter the tasks here https://github.com/web-platform-tests/wpt.fyi/blob/56c72b7c2eeccb4c6433833e3d790131759d1eec/api/taskcluster/webhook.go#L289

**Comment by stephenmcgruer** [comment]
That also sounds good. In either case, sounds like we have a plan for wpt.fyi so will make sure to incorporate that once we start working on this.

**Comment by stephenmcgruer** [comment]
@ziransun as we discussed earlier today, I took a brief look at what would probably be needed here and wrote up some (very) high level steps. Feel free to read them through, look at the files and explore the flow, then I'm happy to answer questions (it is non-trivial!) :)

`tools/ci/tc/tasks/test.yml` is a yaml file that is used by our 'decision graph' code for TaskCluster to decide what to run. It defines the configurations that we run, what triggers they run on, etc. I believe we could modify entirely this file to achieve regular TaskCluster runs, something like:

1. Add a new trigger for a daily run of chrome dev :
```
tasks:
  # Run full suites on push
  - $map:
...
      do:
        $map:
          for:
...
            - vars:
                browser: chrome
                channel: dev
              use:
                - trigger-daily
```
2. Also add some sort of `var` for python 3, which is set for the above run
3. Update the 'wpt-run' definition so that it adds 'python3' to the task name somehow, and passes that to ./tools/ci/taskcluster-run.py (just add `--py3` I think, it passes args directly to 'wpt run' if I read the code correctly).

The above won't use Python3 for everything (the actual start-point is `.taskcluster.yml`, which runs `./wpt tc-decision` which won't be running under Python3), but it would be a start and gets us what we are interested in (a comparison of py2 to py3 via wpt.fyi) without too much effort.

**Comment by ziransun** [comment]
@stephenmcgruer Thanks very much! I will start looking into it.

**Comment by stephenmcgruer** [comment]
To be revisited once we're closer to switching to Python 3 as our 'main' Python. For now we will rely on periodical trigger runs to get comparative results.
